DefaultModuleService: change persisting of items

stelfrich opened the pull request:
Disable the persisting of a `ModuleItem`s if an `initialize()` method is set and is not empty. This way, `initialize()` can overwrite the value of a `ModuleItem` even if `persist=true` is assumed implicitly.

See:
- https://github.com/imagej/imagej-legacy/pull/137#issuecomment-200543860

stelfrich committed:
DefaultModuleService: change persisting of items

Disable the persisting of a ModuleItems if an initialize() method is set
and is not empty. This way, initialize() can overwrite the value of a
ModuleItem even if persist=true is assumed implicitly.

See:
* https://github.com/imagej/imagej-legacy/pull/137#issuecomment-200543860

ctrueden commented:
Did this work well in your tests?

stelfrich commented:
It works as I expect it to work after resetting the Prefs. That leaves us with two issues:

1. If a `ModuleItem` has been persisted previously, this value will still overwrite `initialize()`
2. With this change, explicitly setting `persist=true` for a `ModuleItem` that also has `initializer` set, will disable persisting (that's what we have expected)

I couldn't find any other odd behavior. But as you have already mentioned, we might be breaking someone else's code with this change.

stelfrich committed:
Parameter: update initializer() documentation

stelfrich commented:
> With this change, explicitly setting `persist=true` for a `ModuleItem` that also has `initializer` set, will disable persisting (that's what we have expected)

I have added clarifying documentation in `Parameter`.

stelfrich committed:
Add test for persistance behavior of ModuleItems

ctrueden commented on src/test/java/org/scijava/module/ModuleServiceTest.java:
```diff
@@ -34,8 +34,12 @@
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertSame;
 
+import java.security.GeneralSecurityException;
```
???

ctrueden commented on src/test/java/org/scijava/module/ModuleServiceTest.java:
```diff
@@ -34,8 +34,12 @@
 
+import java.security.GeneralSecurityException;
+
+import org.junit.Assert;
```
The typical way to do it is:
```
import static org.junit.Assert.assertEquals;
```
And then below, just `assertEquals` rather than `Assert.assertEquals`.

stelfrich committed:
ModuleServiceText: clean up

stelfrich commented:
Thanks for the comments (and being so patient), @ctrueden! They have been addressed.

ctrueden commented:
I gave this a little more thought, and here is what I would like to do:

* Change `persist` to a ternary enum: YES, NO and DEFAULT or some such. We can have DEFAULT behave as you have coded here—and in fact, tweak the persistence heuristic as needed for what we deem is most reasonable going forward. If someone says YES, they know better, and want it saved. And similarly, NO means NO. :wink:

I would like to include this update as part of the SJC3 release, due within the next few months. There is an `sjc3` integration branch containing all such changes, where the above changes should go.

If you have time to tackle it, great. Or if not, I will take care of it before SJC3 is released.

stelfrich committed:
Add ItemPersistence to denote persistence behavior

Introduces the ItemPersistence enum (YES, NO, DEFAULT). YES and NO
overwrite additional checks (i.e. precedence of an initializer, value
is default value). In case of DEFAULT, the value is persisted only if
it is not the default value and no initializer is set.

stelfrich commented:
> If you have time to tackle it, great.

I did, @ctrueden. But that invalidates all `Command`s with bytecode of the previous `Parameter` implementation, right? How would I test this implementation in an ImageJ2 installation - I am a bit lost here :confused:

ctrueden commented:
FYI, my plan is to merge this into the `sjc3` branch. Will do soon.

ctrueden commented:
Just a quick update: I looked into doing this earlier in January, but the `sjc3` branch of scijava-common is quite out of date. It needs a rebase over the latest scijava-common. In particular, all the logging updates are now incompatible with the changes on master as well as the `log-revamp` branch.

So it will be a while longer before I have time to get all this unified. Probably April. I am anxious to get SJC3 finished and merged, but it's going to be a substantial endeavor.
